Bitbucket Server Root Context

nwinkler opened the pull request:
I've finally found the time to take a look at the optional root context for self-hosted instances of Bitbucket Server. In our example, the Bitbucket server instance is not located at

https://bitbucket.example.com/scm/ppp/test-repo.git

but has an added `/git/` before the `scm` part:

https://bitbucket.example.com/git/scm/ppp/test-repo.git

We had discussed this a while back in https://github.com/paulirish/git-open/pull/83#issuecomment-309355515, but haven't had an implementation that took anything before the `scm` part into account.

I've refactored the Bitbucket logic to match the logic that was added for TFS, where the check for the `scm` part is now down from the back of the URL (the third to last item in the URL), instead of checking from the front.

No changes in the existing functionality, it just added support for self-hosted instances of Bitbucket Server that use an additional root context in the server deployment. Any optional path elements found before the `scm` match are now added to the final URL as part of the path transformation for Bitbucket Server. Please see the new unit tests for additional examples.

derimagia commented:
The idea for these type of things (which could exist in gitlab or github) is to use the git config to set this ala #92 - so not a bitbucket specific issue. Did you try this yet? We could merge the fix for bitbucket only although I think using the config here makes more sense.

nwinkler commented:
Good points, but this is a different use case. For the GitLab case, you're trying to change a _remote_ URL to a completely different URL, e.g. (from the unit tests):

* Remote URL: `git.example.com`
* Browser URL: `gitlab.example.com`

This means that the remote URL does not have all of the information required for creating the browser URL, so you need to source it in from somewhere else (`git config` in this case).

In the Bitbucket Server case that I added now, all of the information is in the _remote_ URL, it just wasn't used before:

* Remote URL: `bitbucket.example.com/really/long/root/context/scm/ppp/test-repo.git`
* Browser URL: `bitbucket.example.com/really/long/root/context/projects/ppp/repos/test-repo`

Before my change, the `really/long/root/context` part was simply not used. The previous version of the code was checking for the `scm` element in position 0 of the path, whereas I have changed it now to look at the third to last position, allowing me to use the parts that come before `scm` for constructing the browser URL.

In this case, no external information needs to be sourced in from `git config`, since the remote URL already has all of the information.

I was able to remove one further `if` check from the code to clean it up a bit.

Does this make sense?

4U6U57 commented on test/git-open.bats:
```diff
@@ -269,6 +269,35 @@ setup() {
 }
 
+
+@test "bitbucket server with different root context" {
```
**TYPO:** Please add colon here (`bitbucket: `) so the CI output looks nice. Same on the following unit test.

nwinkler replied:
Ah yes, thanks for noticing this. I also saw that some of the basic Bitbucket Server tests were duplicated in an earlier commit - I'll fix this at the same time as the naming...

nwinkler commented:
I've fixed the naming of the test cases, and I've removed some duplicated tests for Bitbucket Server functionality as well. It looks like they were introduced as part of a previous commit/merge when the test suite was refactored.

derimagia commented on test/git-open.bats:
```diff
@@ -269,6 +269,35 @@ setup() {
+  assert_output "https://mybb.domain.com/root/context/projects/~first.last/repos/rrr/browse?at=develop" ||
+    assert_output "https://mybb.domain.com/root/context/projects/~first.last/repos/rrr/browse?at=refs%2Fheads%2Fdevelop" ||
+    assert_output "https://mybb.domain.com/root/context/projects/~first.last/repos/rrr/browse?at=refs/heads/develop"
+
```
#nitpick - remove the space?

4U6U57 replied:
Saw this too, but it's indentation for the chained `||`.

We could possibly reduce duplicate strings by making use of `assert_output --regexp`, although it would be less readable. Like so:

```bash
assert_output --regexp "https://mybb.domain.com/root/context/projects/~first.last/repos/rrr/browse\?at=(refs(/|%2F)heads(/|%2F))?develop"
```

nwinkler replied:
I've simply copied this part from the other existing unit tests, including the `||` chaining and the indentation.

Not a huge fan of the regexp check, due to readability...

derimagia commented:
I see, so you want to change the url before the code here hits for bitbucket: https://github.com/paulirish/git-open/blob/8caa9ad1da1a905b9f7db22ae7085a086f719608/git-open#L130-L134. So the other service we do custom work in is Team Foundation Server and it looks like this is similar to what that does. I'm good with this then 👍 .

nwinkler commented:
Sorry to keep coming back to this - but is anything preventing this from getting merged? Let me know and I'll look into making any additional changes if required...

derimagia commented:
👍

nwinkler commented:
Thanks for merging!
